mount/mount_darwin.go

```diff
@@ -0,0 +1,58 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package mount
+
+import (
+	"os"
+
+	"github.com/pkg/errors"
+)
+
+var (
+	// ErrNotImplementOnUnix is returned for methods that are not implemented
+	ErrNotImplementOnUnix = errors.New("not implemented under darwin")
+)
+
+// Mount to the provided target path.
+//
+// Use symlink instead of union mount on darwin
+func (m *Mount) Mount(target string) error {
+	if m.Type != "bind" {
+		return errors.Errorf("invalid mount type: '%s'", m.Type)
+	}
+
+	if err := os.Remove(target); err != nil {
+		return err
+	}
+	if err := os.Symlink(m.Source, target); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Unmount is just a rm of symlink
+func Unmount(mount string, flags int) error {
+	os.RemoveAll(mount)
+	return os.Mkdir(mount, 0755)
+}
+
+// UnmountAll is just a rm of symlink
+func UnmountAll(mount string, flags int) error {
+	os.RemoveAll(mount)
+	return os.Mkdir(mount, 0755)
+}
```

Comment:
Is it possible to move the symlink hack from the mount package to the naive snapshotter? 

macOS is basically BSD-flavored Unix and has [mount(2)](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/mount.2.html). The current implementation may work for the snapshotter but not for other use cases.

Reply:
> Is it possible to move the symlink hack from the mount package to the naive snapshotter?

The hack is intended to emulate bind mount of Linux, and the snapshotter side is _not_ aware of  how it is mounted (If I understand correctly).
I understand your comment that `Unmount` may possibly unmount unintended target (and delete files/directories).

What about adding a check if the target is a symbolic link or not (like below) ?


```suggestion
func Unmount(target string, flags int) error {
	if isSymlink(target) {
		return unmountSymlink(target, 0755)
	}

	return errors.Wrapf(unix.EBUSY, "failed to unmount target %s", target)
}
```

Reply:
You are right. The snapshotter doesn't call Mount by itself.

I can agree that having the symlink check is safer than the current revision. Can you also handle non-symlink cases?

Reply:
thanks for the comment, 

> I can agree that having the symlink check is safer than the current revision. Can you also handle non-symlink cases?

let me try.

Reply:
i have updated (force-pushed) the branch adding checks on unmount while calling unmount(2) syscall in non-symlink cases.  let me know how you think.

Reply:
no, the updated patch doesn't work..   let me try again.

Reply:
I have updated again.

because unmount is called multiple times on a single target, like during cleanup process of shim instances, it should handle properly after second unmount (but it wasn't in my previous patch).  I fixed it.